sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -261,7 +261,7 @@ final class Decimal extends Ordered[Decimal] with Serializable {
 
   def - (that: Decimal): Decimal = Decimal(toBigDecimal - that.toBigDecimal)
 
-  def * (that: Decimal): Decimal = Decimal(toBigDecimal * that.toBigDecimal)
+  def * (that: Decimal): Decimal = Decimal(toJavaBigDecimal.multiply(that.toJavaBigDecimal))
```

Comment:
This is kind of workaround, didn't fix the root cause. 

The root cause should be in `toBigDecimal`, which doesn't carry on the information about precision to BigDecimal. Could you fix that?